File: py/algorithm/pca_model.py
import numpy as np
import py.algebra.albebra_util as au

class PcaModel():

    def train(self, samples, dimension):
        n = len(samples)
        m = samples[0].shape[0]
        input = np.zeros((n, m))
        for i in range(n):
            for j in range(m):
                input[i][j] = samples[i][j, 0]
        # Step 1. Calculate the mean vector u
        self.u = au.mean(input, 0)

        # Step 2. Calculate the scattering matrix S
        self.s = np.zeros((m, m))
        for i in range(n):
            x = input[i]
            temp = np.dot(np.subtract(x, self.u), np.transpose(np.subtract(x, self.u)))
            self.s = np.add(self.s, temp)

        # Step 3. Calculate the mapping matrix A
        self.a = np.zeros((m, dimension))
        eigenvalues, eigenvectors = au.eigen(self.s)
        for i in range(dimension):
            self.a[:, i] = eigenvectors[:, i]

    def encode(self, x):
        z = np.dot(np.transpose(self.a), x)
        return z

File: src/main/java/jm/app/algorithm/PcaModel.java
package jm.app.algorithm;

import jm.app.algebra.Matrix;

import static jm.app.algebra.AdvancedAlgebraUtil.*;

public class PcaModel {

    // the mean vector u
    private Matrix u;

    // the scattering matrix S
    private Matrix s;

    // the mapping matrix A
    private Matrix a;

    public void train(Matrix[] x, int dimension) {
        int n = x.length;
        int m = x[0].getRowNum();
        // Step 1. Calculate the mean vector
        Matrix input = new Matrix(n, m);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                input.setValue(i, j, x[i].getValue(j, 0));
            }
        }
        u = transpose(mean(input, 0));

        // Step 2. Calculate the scattering matrix
        s = new Matrix(m, m);
        for (int i = 0; i < n; i++) {
            Matrix temp = multiply(subtract(x[i], u), transpose(subtract(x[i], u)));
            s = add(s, temp);
        }

        // Step 3. Calculate the mapping matrix A
        Matrix[] eigenvaluesAndEigenVectors = eigen(s);
        Matrix eigenValues = eigenvaluesAndEigenVectors[0];
        Matrix eigenVectors = eigenvaluesAndEigenVectors[1];
        a = new Matrix(m, dimension);
        for (int i = 0; i < dimension; i++) {
            Matrix eigenVector = getColumnVector(eigenVectors, i);
            Matrix unitizedEigenVector = unitize(eigenVector);
            a = setColumnVector(a, i, unitizedEigenVector);
        }
    }

    public Matrix encode(Matrix x) {
        Matrix z = multiply(transpose(a), x);
        return z;
    }

}
